### PR TITLE
[WIP] Prototype nested sets for regulations.

### LINKED ISSUE #2: Use the nested set model for storage
Currently, when storing the regulation tree, we store _each_ subtree, keyed by label. While this makes processing very simple, it leads to a great deal of redundancy and can be quite slow when importing the tree (as the structure must be walked and each subtree inserted).

An alternative is the [nested set model](http://en.wikipedia.org/wiki/Nested_set_model), which stores each node once but makes grabbing subtrees (equivalent to subsets) painless. There's even a [few](https://github.com/django-mptt/django-mptt) [implementations](https://pypi.python.org/pypi/django-treebeard) in django.

### PR DESCRIPTION
**Pull request by jmcarp**
This is incomplete, but I thought I'd put it out for feedback before finishing. Tasks:
- [x] Use nested sets to store regulation hierarchy
- [x] Avoid calling pseudo-private methods like `_transform`
- [x] Restore `bulk_put` behavior if possible
- [ ] Update unit tests

More specifically, it wound up being simpler to cast dicts to ORM records in the regulation write view, rather than transforming internally in `DMRegulation`. This would break the elastic data model without further changes, so that will have to get worked out before this is done.

ping @cmc333333 

[Resolves #2]

### CODE REVIEW

**Comment by jmcarp** [comment]
Another note--I pushed up an attempt at combining mptt with bulk inserts. Since django doesn't support compound columns (boo), I changed the pk of the `Regulation` model to a text column and automatically set it to the concatenated version and label. If regs have predictable pks, then we don't lose parent-child relationships on `bulk_create`.

File: regcore/db/django_models.py [hunk 1]
```diff
@@ -1,46 +1,86 @@
 """Each of the data structures relevant to the API (regulations, notices,
 etc.), implemented using Django models"""
+import collections
+
 from django.core.exceptions import ObjectDoesNotExist
 
 from regcore.models import Diff, Layer, Notice, Regulation
 
 
+def get_tree(regs):
```
Comment (cmc333333) [inline regcore/db/django_models.py]: Minor naming thing, but "tree" is a very overloaded term in the regs code. It looks like you're creating a map of child nodes by parent label, instead?

File: regcore/db/django_models.py [hunk 2]
```diff
@@ -1,46 +1,86 @@
...
+            title=reg.get('title', ''),
+            node_type=reg['node_type'],
+            root=(len(reg['label']) == 1),
+        )
+
+    def _get_id(self, reg, version):
+        return '{}-{}'.format(version, '-'.join(reg['label']))
 
     def bulk_put(self, regs, version, root_label):
         """Store all reg objects"""
```
Comment (cmc333333) [inline regcore/db/django_models.py:73]: If we're changing the signature here (regs used to be a list of dicts), let's update the docstring to make it clear what's going on.

File: regcore/db/django_models.py [hunk 3]
```diff
@@ -1,46 +1,86 @@
...
+
+        # Bulk save children with placeholder tree values
+        for row in rows[1:]:
+            row.tree_id = rows[0].tree_id
+            row.level, row.lft, row.rght = 1, 1, 1
+        with Regulation.objects.disable_mptt_updates():
+            Regulation.objects.bulk_create(rows[1:], batch_size=50)
+
+        # Rebuild partial tree from root node
+        Regulation.objects.partial_rebuild(rows[0].tree_id)
```
Comment (cmc333333) [inline regcore/db/django_models.py]: I'm not quite sure what's going on here -- it _looks_ like you're trying to work around some issues with `mptt`. I'd think that lib would be well versed in bulk inserts, however. If we need a workaround, would you mind adding comments to explain why?

File: regcore/models.py [hunk 4]
```diff
@@ -1,21 +1,27 @@
...
     node_type = models.SlugField(max_length=10)
-    children = CompressedJSONField()
     root = models.BooleanField(default=False, db_index=True)
 
     class Meta:
         index_together = (('version', 'label_string'),)
         unique_together = (('version', 'label_string'),)
 
+    class MPTTMeta:
+        order_insertion_by = ['label_string']
```
Comment (cmc333333) [inline regcore/models.py]: I haven't looked too deeply into this, but I'm guessing this is how mptt orders the children of a node. Unfortunately, `label_string` is not a good mechanism for ordering. Consider:
- `111-2`, `111-11`
- `111-1-a-vi`, `111-1-a-ix`
- `111-1-p45`, `111-1-p123456784` (a definition)

In each case, a simple string comparison will fail.

File: regcore/db/django_models.py (hunk 3 above)
Comment (cmc333333) [inline regcore/db/django_models.py]: Related: feel free to remove `bulk_put` in favor of `put` if that makes more sense. It'd be easy to move the list-ifying step into the ElasticSearch storage.

Comment (jmcarp) [inline regcore/db/django_models.py]: Re. workaround: yes, exactly. Since bulk updates bypass a lot of the normal behavior (calling `save`, emitting events, updating saved objects, etc.), django-mptt doesn't handle them for free. The recommendation in the docs and issues is to disable mptt updates, then explicitly rebuild the full or partial tree after bulk operations. I haven't gotten too far into the mptt internals yet, but as a quick explanation, none of this happens on `bulk_create`: https://github.com/django-mptt/django-mptt/blob/cf857138b2cb5eebc465c912c86e804ed975dd99/mptt/models.py#L807-L996.

Does this make sense? I may still be missing something.

File: regcore/models.py (hunk 4 above)
Comment (jmcarp) [inline regcore/models.py]: Ack, yes. Maybe it would be simplest to explicitly encode ordering in the ORM--we could `enumerate` the children in the write view, save the index as `position` or something, and use that column for insertion order. Unless there's an adequate key or set of keys already on this model.

File: regcore/db/django_models.py (hunk 1 above)
Comment (jmcarp) [inline regcore/db/django_models.py]: Agree--will fix.

File: regcore/db/django_models.py (hunk 2 above)
Comment (jmcarp) [inline regcore/db/django_models.py:73]: Sorry, I changed the signature a few times here--as of the latest commit, `regs` is a list of `(node, parent)` tuples, where `node` and `parent` are dicts. Which is more similar to the current interface, and doesn't require calling private methods outside their class. Anyway, agree--I'll go through and update all relevant docstrings once we're happy with the patch.

**Comment by jmcarp** [comment]
A few more things about this:
- I ran some less-lazy benchmarks, and the MPTT patch comes out slower than master for both reads and writes. Some of that happens because even with the `save=False` flag, django-mptt runs a bunch of queries against the database. Even when I comment out those queries, inserts are still slightly slower than on master. Meanwhile, reads from the API go from very fast (~32ms for `/regulation/555/03-6573`) to just slightly slower (~40ms). It's a big proportional difference but a tiny absolute difference, and in practice we can cache aggressively using both api.data.gov and cached partials in -site.
- Switching to django-mptt reduces disk usage on the `regcore_regulation` table by roughly 75%, but it's not clear that we care about that much--storage is cheap, and overall storage use seems pretty low either way.
- Now that I think about it, it's not clear that we need a relational database here in the first place, or that normalization is a good thing. The API is effectively a key-value store--we index regulations by (version, label), layers by (name, version, label), etc., and read requests just look up a stored value by keys. We aren't exposing other filters, sorts, joins, aggregations, etc. As a thought experiment, I wonder if we couldn't just enumerate all possible lookups, write each one to a file, and serve the whole API statically via nginx, or S3, or a CDN. Not that there's enough data for hosting costs to be a concern, but S3 is way cheaper than equivalent space on RDS or cloud.gov.

In other words, I overthought this issue, and now I'm not sure whether there's a worthwhile performance benefit to be had here. Unless I'm still using django-mptt inefficiently, which is totally possible.

**Comment by cmc333333** [comment]
Pardon the rambling, but here goes anyway.

Some context: the major impetus for this was finding 40 CFR 52, which is roughly as large as all of the other regs we've seen combined. We were already pushing the limit of what RDMS storage can happily accept (MySQL had a 1M message size at CFPB, iirc; we also ran the risk of the HTTP interface timing out). A reg that large surely would trigger both problems.

Re S3/file storage: you are correct, the API has largely been a key-value store. It initially grew out of the need to search (hence the ElasticSearch storage backend) but was jammed into an RDMS due to technical policies. That said, I think there's a lot of relational data that we haven't made use of; it'd be super handy to know the most cited paragraph or section, for example, which is not easily answerable due to the key-value architecture. I'd prefer we move towards making the data easier to query than locking it up in binary blobs.

On static content in general, we've talked a lot about making the eRegs UI a static site generator. I think this would be a good direction to move as it's much easier to host and hand off to clients. We do have some dynamic components of the UI (particularly search), however, so we'd need to either completely disable these features or have some sort of mixed static-dynamic hybrid. In any event, that's a can of worms.

Unfortunately, the API is pulled in two rather different directions. On one hand, we want to make it as flexible as possible, making a datastore that's easy to query; we want the API itself to be a product. On the other hand, there is only one consumer (the eRegs UI) and that has really molded the API's development to match its needs. This leads us to a rather annoying spot where we've implemented an API-first system, yet spent almost no time making a general purpose API. This has lead to duplication -- consider the parser's local "index", which should probably be backed by this API instead.

All of this is to say, I'm not sure the best approach. I think the storage savings are actually a big win in light of larger regulations, but that need isn't super pressing. I'd be content to close this PR and say "now we know the strategy is viable" and let sleeping dogs lie until we need to disturb them.

**Comment by jmcarp** [comment]
Context much appreciated!

I don't mean to seriously propose switching from *SQL to files--even if we don't wind up using more RDBMS features eventually, I don't see much value in changing. Especially if we're thinking about or planning to make the API more flexible.

Have we parsed 40 CFR 52? I think it would be interesting to see how a much larger regulation works under this patch compared to master. Either way, I'm thinking we might run into problems retrieving the root node and all its descendants in a single request. Like you said, it could be too big to fit into a single row under master, and even after we normalize the data, calling `DMRegulations#get` is going to fetch and serialize a huge number of descendants.

I'm wondering if we'd want to normalize layers too. Instead of storing layers in big blobs, we should be able to find all layers for a given regulation and its descendants using something like `Layer.objects.filter(regulation__lft__gte=reg.lft, regulation__rght_lte=reg.rght, ...)`. I was messing with this briefly, and it seems to work fine, although lookups are also a bit slower than now--which is expected, since we're now doing a join and deserializing many layers instead of loading a single blob with all the relevant information already assembled.

Anyway, I'm happy to get this patch into a merge-able state now, or to defer until 40 CFR 52 is ready to mess with. What do you think @cmc333333?

**Comment by cmc333333** [comment]
> Have we parsed 40 CFR 52?

Unfortunately no: 18F/regulations-parser#194. It might make sense to construct this for testing purposes

> I'm wondering if we'd want to normalize layers too

I agree. I think we would have two basic layer types (maybe splitting into two models): an "inline" type, which targets a specific chunk of the text and a "meta" type, which applies to a while paragraph/section. In any event, I think normalizing the tree could be a great first step.

>  I'm happy to get this patch into a merge-able state now

Let's do it! There's no huge rush here, if something more important comes up, but I think this would still be worthwhile. One missing piece thus far is a data migration. This should be pretty simple -- delete the non-root rows then resave the roots.

**Comment by jmcarp** [comment]
Cool, I'll look over #36 and #37 and then get back to this.

**Comment by jmcarp** [comment]
Replaced by #38.
